## Some ideas for your final project


- *Batch effect correction*
    - Leek, J. T. & Storey, J. D. [Capturing Heterogeneity in Gene Expression Studies by Surrogate Variable Analysis](http://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.0030161). PLoS Genet 3, e161 (2007). [TAKEN: Gloria and Marea]

- *RNA-seq*
    - Anders, S., Reyes, A. & Huber, W. [Detecting differential usage of exons from RNA-seq data](http://www.ncbi.nlm.nih.gov/pubmed/22722343). Genome Res. 22, 2008-2017 (2012). 
    - Frazee, A. C., Sabunciyan, S., Hansen, K. D., Irizarry, R. A. & Leek, J. T. [Differential expression analysis of RNA-seq data at single-base resolution](http://biostatistics.oxfordjournals.org/content/early/2014/01/06/biostatistics.kxt053.full). Biostatistics kxt053 (2014). doi:10.1093/biostatistics/kxt053. [TAKEN: Yatong, Xiaowen and Kelsey]
    - Risso, D., Ngai, J., Speed, T. P. & Dudoit, S. [Normalization of RNA-seq data using factor analysis of control genes or samples](http://www.nature.com/nbt/journal/v32/n9/full/nbt.2931.html). Nature Biotechnology 32, 896–902 (2014). [TAKEN: Ben and Aaron]
    
    
- *Single-cell gene-expression*
    - McDavid, A. et al. [Data exploration, quality control and testing in single-cell qPCR-based gene expression experiments](http://www.ncbi.nlm.nih.gov/pmc/articles/PMC3570210/). Bioinformatics 29, 461-467 (2013).
    - Shalek, A. K. et al. [Single-cell transcriptomics reveals bimodality in expression and splicing in immune cells](http://nrs.harvard.edu/urn-3:HUL.InstRepos:11879192). Nature 498, 236-240 (2013). 
    - Trapnell, C. et al. [The dynamics and regulators of cell fate decisions are revealed by pseudotemporal ordering of single cells](http://www.nature.com/nbt/journal/v32/n4/full/nbt.2859.html). Nature Biotechnology 32, 381–386 (2014).

- *Gene expression deconvolution*
    - Shen-Orr, S. S. et al. [Cell type-specific gene expression differences in complex tissues](http://www.ncbi.nlm.nih.gov/pmc/articles/PMC3699332/). Nat. Methods 7, 287-289 (2010). 
    - Zhong, Y. & Liu, Z. [Gene expression deconvolution in linear space](http://www.nature.com/nmeth/journal/v9/n1/full/nmeth.1830.html). Nat. Methods 9, 8–9– author reply 9 (2012). [TAKEN: Benjamin & Sara]

- *Methylation*
    - Hansen, K. D., Langmead, B. & Irizarry, R. A. [BSmooth: from whole genome bisulfite sequencing reads to differentially methylated regions](http://www.ncbi.nlm.nih.gov/pmc/articles/PMC3491411/). Genome Biol. (2012).
   
- *Flow cytometry*
    - Finak, G. et al. [OpenCyto: an open source infrastructure for scalable, robust, reproducible, and automated, end-to-end flow cytometry data analysis](http://www.ncbi.nlm.nih.gov/pmc/articles/PMC4148203/). PLoS Comput. Biol. 10, e1003806 (2014).
    - Amir, E.-A. D. et al. [viSNE enables visualization of high dimensional single-cell data and reveals phenotypic heterogeneity of leukemia](http://www.ncbi.nlm.nih.gov/pmc/articles/PMC4076922/). Nature Biotechnology 31, 545–552 (2013).

## What you will have to do
1. Read and understand the paper
2. Reproduce some of the analysis
3. Summarize the paper with the analysis in reproducible presentation (a .Rmd file)
4. Present your work to the rest of the class
